### PR TITLE
User/kcgroulx/mcb/nucleo testing

### PR DESCRIPTION
**Pull request by kcgroulx**
First time doing a pull request 💀

### CODE REVIEW

File: MCB-Nucleo/Core/Src/MCB.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/MCB.c:1]: Try to keep convention with file naming, here `mcb.c`

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: State of charge would be more clear as SOC. Make sure to update all related variables/functions

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Using cube to configure threads is simple and good, but after auto generation remember to not create the default thread. This would use resources for no reason. I see that this is blinking the onboard led, but change the task or make a new one for status leds.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: You've already addressed this with other parts of code like 'ADC_DEADZONE' but as a further note, the use of macros in firmware is often beneficial for clarity. I would stop and question everytime there is a 'magic number' in the code. A magic number is any raw value that is arbitrary. Numbers aren't arbitrary to the programmer when coding, but in a couple months they can become unclear when everyone starts to question "why is this number this value". This is a concept of designing for longevity and quality code because it's all to common that we have no clue what past members were doing in their code.

Here the use of a comment is one solution. Another solution is to use a macro/variable with a clear name. This removes the need to comment code and directly clarifies things. For example, `SOC_MAX_THRESHOLD ` and `SOC_MIN_THRESHOLD` could be macros. This is especially helpful if we change the threshold to be something less obvious like 97.

^This is actually added as a macro in `MCB.h` and it is equal to 90. I would resolve this inconsistency.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Typically interrupt related code would go in `stm32f1xx_it.h` for STM based projects. This is a thing on clarity and putting code in suitable locations. This is code related to the rtos but not that directly. It may be more suitable in the interrupt file or `gpio.c` file.

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Inc/MCB.h]: Adding to magic number discussion in `freertos.c`, I would add a comment on all of these raw values. This would not be clear for a developer who was reading the codebase from new. What is the number representing - maybe comment it's for a specific type of pot. It might be even better to clarify 'ADC' with something more specific because there can be multiple ADC readings.

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Inc/MCB.h]: Same discussion on magic numbers, I would add a comment that just says milliseconds or use `SEND_MOTOR_COMMAND_DELAY_MS`. This may seem useless, but it takes practically no time and saves future you/future developers a short bit of time for code comprehension. Using more characters in a variable name for clarity is rarely a bad thing.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Machi typo?

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Nit: styling braces

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Are all/any of these threads "starting" something? I would question the naming here.

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Inc/MCB.h]: Typo

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Outside code. Make sure this is specced that the cruise button also turns off cruise - if that is the desired behavior.

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Inc/MCB.h]: What are the units and how did these get chosen?

File: MCB-Nucleo/Core/Src/main.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/main.c]: Is this supposed to be in `main.c`?

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Make sure to add more details to the documentation. This doesn't really say anything.

For example: "A thread that outputs motor drive commands for all states of the board"

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: Please finalize the mcb specs doc. We discussed using regen braking in park mode to stop the car from rolling. If that is or is not being implemented it should be specified. Here that would mean max current.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/freertos.c]: This may benefit from comments about the hierarchical structure of the conditional which achieves the functionality/edge case behavior specced out.

File: MCB-Nucleo/Core/Src/MCB.c
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Src/MCB.c]: Return not needed.

**Review by MatthewChow03 — CHANGES_REQUESTED** [review]
Great work, keep it up! Here are some notes and the rest are inline with the code. There are specific things to change and the rest are tips for PRs in general. We haven't discussed the development process/workflow at Solar so this will be something we can make clearer in the future. Also, there are a lot of comments because this is a large PR.
- Make the PR description informative and to the point. This should be a helpful comment of what the PR is for the people reviewing. Context of the code is key. Consider what info would benefit your reviewer depending on the context.
-  In the future, we will aim to PR chunks of code together - individual features that are related. Here there is a lot going on since the whole project is coming in 1 PR. For example, a PR could be implementing cruise control or adding can transmission to the mcb. This iterative process allows for continuous development and feedback that improves code quality because individual chunks of code get reviewed in isolation and build up iteratively.
- Pull master before pushing. The PR is merging into master so it can't conflict with everything there. Put the `mcb` project in the `components` folder as we discussed. You will notice that in the files changed, there are a lot of related files. This is because the bms projects are deleted in your PR. In general, make sure you check what files changed and know that the code you are committing/submitting for PR is exactly what you want to be pushing.
- Comments on the PR are often suggestions while some are todos. Make sure to take them as questions and consider how they could improve the code quality. Take a skim through all the comments before getting to coding.
- Note that testing is for the actual state machine logic, I'm not covering that in this PR

**Comment by MatthewChow03** [comment]
Also, these will appear as inline comments in VS Code. Not sure about STM32IDE.

**Comment by a2k-hanlon** [comment]
Lots of good work here Kyle, and excellent feedback Matthew!

I haven't read through the code here like Matthew has, just a couple of comments on the organization of branches for the MCB and what is going on here with the BMS firmware files.

1. I think you guys need to consider where this branch stands with respect to the branch `mcb_main`. As I understand, that has been the target branch for all MCB work so far. It looks like this branch is based on that one. Now, my feeling is that `mcb_main` should be merged to master and master should be the target branch for changes to the firmware for all the devices in this repo. It would probably be best to do that merge explictly before merging this branch to master. If you merge this to master without doing that, you'll have effectively merge `mcb_main` stuff plus your stuff, which is messy version control. Also see Matthew's comment about locating MCB code in the `components/mcb` folder.

2. Matthew already mentioned this but I'd like to second that you must not merge this with the deletion of all of the BMS firmware files. Please revert that and ensure that no changes to BMS files show up in the differences for your PR (that's the "Files Changed" tab of this page).

Please don't take this as criticism of what you've done here, it isn't! Just hoping to straighten out some of the version control cleanliness wrinkles we've got going on. This definitely follows the thread of Matthew's comment about us lacking established and consistent guidelines for our development workflow.

**Comment by MatthewChow03** [comment]
> 1. I think you guys need to consider where this branch stands with respect to the branch `mcb_main`. As I understand, that has been the target branch for all MCB work so far. It looks like this branch is based on that one. Now, my feeling is that `mcb_main` should be merged to master and master should be the target branch for changes to the firmware for all the devices in this repo. It would probably be best to do that merge explictly before merging this branch to master. If you merge this to master without doing that, you'll have effectively merge `mcb_main` stuff plus your stuff, which is messy version control. Also see Matthew's comment about locating MCB code in the `components/mcb` folder.

@a2k-hanlon This branch is the most up to date MCB code and contains everything in the `mcb_main` branch - besides 1 commit which isn't important. I was thinking Kyle would merge this branch into master and then we would delete all the other MCB branches. We would do a similar thing with TEL soon where Claris would merge in her current branch and then we delete all the others since they are subsets of the working branch. We didn't make use of the protected `xxx_main` branches. How would this make the version control more messy? I don't think I see how it's different than merging `mcb_main` and then merging this branch - besides having a larger merge.

**Comment by a2k-hanlon** [comment]
@MatthewChow03 By messy version control I mean this: If this PR is making **distinct** edits to the existing mcb code (Mihir's and your commits) prior to Kyle's commits, while including that existing code in the PR rather than merging it first, you won't clearly see the differences introduced to that existing code in this PR. In the diff, it all just looks like new code. That doesn't follow your "PR chunks of code together" comment which is a comment I definitely agree with. It can be helpful to have PRs with clear, distinct code diffs as records of when and why changes are made. 

With that said, I'm not sure if Kyle's work is a set of distinct changes or a lot of more significant rework since at the moment Kyle has recreated the MCB project in a different folder and the diff is all addition of code for that reason anyway. If Kyle's changes are significant rework, maybe being able to see exactly what Kyle's changes are with respect to the older MCB code may not be important.

In any case, I trust that you guys know best what's going on with the MCB code so seeing a clear diff for Kyle's work may not be relevant. Also, you are right that the end result of merging this PR with the older mcb commits and dropping other mcb branches would be the same as merging mcb_main followed by this. Doing the latter instead would only change what code differences are shown in each of those two pull requests - that's all I wanted you to consider. If just merging this PR, with the entirety of the work on MCB code, works for you that's cool with me!

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: image
From what I understand, the default task cannot be deleted since it handles initializations for freeRTOS. I'm using it as a heartbeat LED for an easier way to see if the code is running. I think it should be fine to delete another task and place its code into the default task.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Added macros for battery full and empty.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Moved to stm32f1xx_it.h

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (kcgroulx) [reply MCB-Nucleo/Core/Inc/MCB.h]: Added comments for defines

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (kcgroulx) [reply MCB-Nucleo/Core/Inc/MCB.h]: Added comments for defines

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Fixed typo

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Fixed, added parentheses

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: FreeRTOS required both a "Task Name" as well as a "Entry Function." 
I used the same naming convention as the defaultTask that auto-generates with the program. 
("defaultTask" as the Task Name and "StartDefaultTask" as the Entry Fuction").
I agree that the use of "start" is not intuitive so I changed it to the same naming convention I use at work.

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (kcgroulx) [reply MCB-Nucleo/Core/Inc/MCB.h]: Fixed typo

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (kcgroulx) [reply MCB-Nucleo/Core/Inc/MCB.h]: Units are m/s, I added a comment to specify this.
Values are placeholders and need to be specified,

File: MCB-Nucleo/Core/Src/main.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/main.c]: Removed it, not being used anymore.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Added basic descriptions for each task

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Changed current value to MAX in PARK and updated specs doc to reflect this.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/freertos.c]: Added a comment explaining the purpose of the hierarchical structure.

File: MCB-Nucleo/Core/Src/MCB.c
Comment (kcgroulx) [reply MCB-Nucleo/Core/Src/MCB.c]: Removed return.

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [reply MCB-Nucleo/Core/Src/freertos.c]: Is this just a cube thing? Could they not be named the same thing? I don't see where the task name shows up in the code

File: MCB-Nucleo/Core/Src/freertos.c
Comment (MatthewChow03) [reply MCB-Nucleo/Core/Src/freertos.c]: Yeah correct. I meant in the code you can comment out the function to create the thread so that it doesn't run. The cube generator is just a tool but you can modify the code to do whatever you want after generation. For the final code of the car you could just delete all the default task code instead of commenting it out.

File: MCB-Nucleo/Core/Src/MCB.c
Comment (MatthewChow03) [reply MCB-Nucleo/Core/Src/MCB.c]: Btw, not sure what people do for collaboration, but if there is nothing to follow up on/discuss I would say it's good to just resolve the conversation, no need to write a comment. After making suggestion as a reviewer, I trust your judgement to implement whatever you think is best based on the suggestion

File: MCB-Nucleo/Core/Inc/MCB.h
Comment (MatthewChow03) [inline MCB-Nucleo/Core/Inc/MCB.h]: Just a note. You can also do UPDATE_FLAGS_TASK_DELAY_MS, just a style option which would make it longer but would remove the need to inspect the macro to read the comment about the macro. This is good as is.

**Pull request closed without merging**
